root@3FG16:/workspace/dpdp_habana/PaddleCustomDevice/backends/intel_hpu/tests/unittests# python test_scatter_hpu.py -v
/usr/local/lib/python3.10/dist-packages/paddle/utils/cpp_extension/extension_utils.py:686: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md
  warnings.warn(warning_message)
I1231 03:29:44.638433 34084 init.cc:235] ENV [CUSTOM_DEVICE_ROOT]=/usr/local/lib/python3.10/dist-packages/paddle_custom_device
I1231 03:29:44.638454 34084 init.cc:144] Try loading custom device libs from: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1231 03:29:44.730544 34084 custom_device.cc:1099] Succeed in loading custom runtime in lib: /usr/local/lib/python3.10/dist-packages/paddle_custom_device/libpaddle-intel-hpu.so
I1231 03:29:44.732095 34084 custom_kernel.cc:63] Succeed in loading 107 custom kernel(s) from loaded lib(s), will be used like native ones.
I1231 03:29:44.732205 34084 init.cc:156] Finished in LoadCustomDevice with libs_path: [/usr/local/lib/python3.10/dist-packages/paddle_custom_device]
I1231 03:29:44.732224 34084 init.cc:241] CustomDevice: intel_hpu, visible devices count: 8
test_check_grad (__main__.test_gather_fp32_no_ow) ... I1231 03:29:47.982786 34084 program_interpreter.cc:243] New Executor is Running.
I1231 03:29:47.993903 34084 interpreter_util.cc:651] Standalone Executor is Used.
ok
test_check_output (__main__.test_gather_fp32_no_ow) ... ok
test_check_grad (__main__.test_gather_fp32_no_ow_2) ... ok
test_check_output (__main__.test_gather_fp32_no_ow_2) ... FAIL
test_check_grad (__main__.test_gather_fp32_no_ow_3) ... ok
test_check_output (__main__.test_gather_fp32_no_ow_3) ... ok
test_check_output (__main__.test_gather_i32_no_ow) ... ok
test_check_output (__main__.test_gather_i32_ow) ... ok
test_check_output (__main__.test_gather_i64_no_ow) ... ok
test_check_output (__main__.test_gather_i64_ow) ... ok

======================================================================
FAIL: test_check_output (__main__.test_gather_fp32_no_ow_2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/dpdp_habana/PaddleCustomDevice/backends/intel_hpu/tests/unittests/test_scatter_hpu.py", line 145, in test_check_output
    self.check_output_with_place(self.place)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2727, in check_output_with_place
    static_checker.check()
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2329, in check
    self.compare_outputs_with_expects()
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2319, in compare_outputs_with_expects
    self.compare_single_output_with_expect(out_name, expect)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2299, in compare_single_output_with_expect
    self._compare_numpy(name, actual_np, expect_np)
  File "/workspace/dpdp_habana/PaddleCustomDevice/python/tests/op_test.py", line 2266, in _compare_numpy
    np.testing.assert_allclose(
  File "/usr/local/lib/python3.10/dist-packages/numpy/testing/_private/utils.py", line 1504, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.10/dist-packages/numpy/testing/_private/utils.py", line 797, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-05, atol=0
Operator (scatter) Output (Out) has diff at Place(intel_hpu:0) in static checker
Mismatched elements: 2 / 6 (33.3%)
Max absolute difference: 1.
Max relative difference: 0.8150946
 x: array([[1.      , 1.      ],
       [0.226851, 0.551315],
       [1.      , 1.      ]], dtype=float32)
 y: array([[1.      , 1.      ],
       [1.226851, 1.551315],
       [1.      , 1.      ]], dtype=float32)

----------------------------------------------------------------------
Ran 10 tests in 3.217s

FAILED (failures=1)

--- a/backends/intel_hpu/tests/unittests/test_scatter_hpu.py
+++ b/backends/intel_hpu/tests/unittests/test_scatter_hpu.py
@@ -122,5 +122,74 @@ class test_gather_i32_no_ow(OpTest):
         self.check_output_with_place(self.place)
 
 
+class test_gather_fp32_no_ow(OpTest):
+    def setUp(self):
+        self.set_hpu()
+        self.op_type = "scatter"
+        self.place = paddle.CustomPlace("intel_hpu", 0)
+        self.python_api = paddle.scatter
+
+        ref_np = np.ones((3, 2)).astype("float32")
+        index_np = np.array([1]).astype("int32")
+        updates_np = np.random.random((1, 2)).astype("float32")
+
+        output_np = np.copy(ref_np)
+        output_np[index_np] = updates_np
+        self.inputs = {"X": ref_np, "Ids": index_np, "Updates": updates_np}
+        self.outputs = {"Out": output_np}
+        self.attrs = {"overwrite": True}
+
+    def set_hpu(self):
+        self.__class__.use_custom_device = True
+
+    def test_check_output(self):
+        self.check_output_with_place(self.place)
+
+    def test_check_grad(self):
+        self.check_grad_with_place(
+            self.place,
+            ["X", "Updates"],
+            "Out",
+            check_dygraph=False,
+            numeric_place=paddle.CPUPlace(),
+        )
+
+
+class test_gather_fp32_no_ow_2(test_gather_fp32_no_ow):
+    def setUp(self):
+        self.set_hpu()
+        self.op_type = "scatter"
+        self.place = paddle.CustomPlace("intel_hpu", 0)
+        self.python_api = paddle.scatter
+
+        ref_np = np.ones((3, 2)).astype("float32")
+        index_np = np.array([1]).astype("int32")
+        updates_np = np.random.random((1, 2)).astype("float32")
+
+        output_np = np.copy(ref_np)
+        output_np[index_np] += updates_np
+        self.inputs = {"X": ref_np, "Ids": index_np, "Updates": updates_np}
+        self.outputs = {"Out": output_np}
+        self.attrs = {"overwrite": False}
+
+
+class test_gather_fp32_no_ow_3(test_gather_fp32_no_ow):
+    def setUp(self):
+        self.set_hpu()
+        self.op_type = "scatter"
+        self.place = paddle.CustomPlace("intel_hpu", 0)
+
+        ref_np = np.ones((3, 2)).astype("float32")
+        index_np = np.array([1, 2]).astype("int32")
+        updates_np = np.random.random((2, 2)).astype("float32")
+
+        output_np = np.copy(ref_np)
+        output_np[1] = updates_np[0]
+        output_np[2] = updates_np[1]
+        self.inputs = {"X": ref_np, "Ids": index_np, "Updates": updates_np}
+        self.outputs = {"Out": output_np}
+        self.attrs = {"overwrite": True}
+
+
 if __name__ == "__main__":
     unittest.main()

